src/connection.js

```diff
@@ -0,0 +1,272 @@
+/*
+ * Copyright 2022, Tony Atkins
+ * Copyright 2011-2020, Colin Clark
+ *
+ *  Licensed under the MIT license, see LICENSE for details.
+ */
+
+(function (fluid) {
+    "use strict";
+    var youme = fluid.registerNamespace("youme");
+
+    /**
+     *
+     * A MIDI Connection represents a connection to a single input or output port.  Unless you are completing the wiring
+     * yourself, you should not use this directly, but should use `youme.connection.input` or `youme.connection.output`
+     * instead.
+     *
+     **/
+    fluid.defaults("youme.connection", {
+        gradeNames: ["fluid.component"],
+
+        openImmediately: false,
+
+        portSpec: {},
+
+        members: {
+            port: false
+        },
+
+        events: {
+            onPortOpen: null,
+            onPortClose: null,
+            onError: null
+        },
+
+        invokers: {
+            open: {
+                funcName: "youme.connection.open",
+                args: ["{that}.port", "{that}.events.onPortOpen.fire", "{that}.events.onError.fire"] // port, onPortOpen, onError
+            },
+
+            close: {
+                funcName: "youme.connection.close",
+                args: ["{that}.port", "{that}.events.onPortClose.fire", "{that}.events.onError.fire"] // port, onPortClose, onError
+            }
+        },
+
+        listeners: {
+            "onCreate.autoOpen": {
+                funcName: "youme.connection.autoOpen",
+                args: [
+                    "{that}.options.openImmediately", "{that}.open"
+                ]
+            },
+
+            "onError.logError": {
+                funcName: "fluid.log",
+                args: [fluid.logLevel.WARN, "{arguments}.0"]
+            },
+
+            "onDestroy.close": "{that}.close"
+        }
+    });
+
+    youme.connection.autoOpen = function (openImmediately, openFn) {
+        if (openImmediately) {
+            openFn();
+        }
+    };
+
+    youme.connection.open = function (port, onPortOpen, onError) {
+        if (port) {
+            var portOpenPromise = port.open();
+            portOpenPromise.then(onPortOpen, onError);
+        }
+        else {
+            onError("Port is missing, cannot open connection.");
+        }
+    };
+
+    youme.connection.close = function (port, onPortClose, onError) {
+        if (port) {
+            var portClosePromise =  port.close();
+            portClosePromise.then(onPortClose, onError);
+        }
+        else {
+            onError("Port is missing, cannot close connection.");
+        }
+    };
+
+    fluid.defaults("youme.connection.input", {
+        gradeNames: ["youme.connection", "youme.messageReceiver"],
+
+        model: {
+            portSpec: {
+                type: "input"
+            }
+        },
+
+        listeners: {
+            "onCreate.startListening": {
```

Comment:
I think I moved this to `onCreate` in a very sloppy fashion, it should always have been bound to `onPortOpen` instead, so that `{that}.open` does indeed offer another chance to bind the listener.

I think I messed this up when I was writing the tests, which failed if I was auto opening, because the port open event triggered during the auto open phase meant that the listener was bound too late for the associated check.  I tidied up the grade and the test.

I'll think about the other pattern, I like this because it matches the previous behaviour, letting authors decide whether to open on create or not.

Reply:
I believe this is now handled cleanly, especially with the modelised open/close stuff.